File: javascript/2023/01/202301.js
//
// Avent of Code 2023 Day 1
// Brian Green
//
//

const fs = require("fs");

// Read data from file and return array
async function readLines(filePath) {
  try {
    const fileContent = await fs.promises.readFile(filePath, "utf-8");
    const lines = fileContent.split("\n");
    return lines;
  } catch (err) {
    console.error("Error reading file:", err);
    return [];
  }
}

// Find the first digit in the string
function findFirstDigit(str) {
  for (let i = 0; i < str.length; i++) {
    const char = str.charAt(i);
    if (!isNaN(parseInt(char))) {
      return parseInt(char);
    }
  }
  return null; // Return null if no digit is found
}

// Find the last digit in the string
function findLastDigit(str) {
  for (let i = str.length - 1; i >= 0; i--) {
    const char = str.charAt(i);
    if (!isNaN(parseInt(char))) {
      return parseInt(char);
    }
  }
  return null;
}

// Find the calibration value for part 1 by adding together the first and last digit
function getCalibrationValue(str) {
  readLines(str).then((lines) => {
    let total = 0;
    //console.log(lines);
    lines.forEach((serialnum) => {
      let calibrate = findFirstDigit(serialnum) * 10 + findLastDigit(serialnum);
      //console.log(calibrate)
      total += calibrate;
    });
    console.log(str, "total=", total);
    return total;
  });
}

// Convert the spelled out numbers into digits and then calculate the calibration value
function getCalValPart2(str) {
  let numName = {
    one: "1",
    two: "2",
    three: "3",
    four: "4",
    five: "5",
    six: "6",
    seven: "7",
    eight: "8",
    nine: "9",
  };
  readLines(str).then((lines) => {
    let total = 0;
    //console.log(lines);
    lines.forEach((serialnum) => {
      Object.keys(numName).forEach((key) => {
        // add extra char at the end of each number found so converting the first overlapping
        // number does not corrupt the next number
        // e.g. oneight -> oneeight do this becomes 18 instead of 1ight
        serialnum = serialnum.replaceAll(key, key + key[key.length - 1]);
        //console.log(key, "new=", serialnum);
      });
      Object.keys(numName).forEach((key) => {
        // replace spelled out numbers with its digit
        serialnum = serialnum.replaceAll(key, numName[key]);
        //console.log("newNew=", serialnum);
      });
      total += findFirstDigit(serialnum) * 10 + findLastDigit(serialnum);
      //console.log("total=", total);
    });
    console.log(str, "(part 2) total=", total);
    return total;
  });
}

getCalibrationValue("example.dat");
getCalValPart2("example2.dat");
getCalibrationValue("data.dat");
getCalValPart2("data.dat");

File: golang/2023/01/202301.go
//
// Advent of Code 2023 Day 1
// Brian Green
//

package main

import (
	"bufio"
	"errors"
	"fmt"
	"os"
	"strings"
)

// ExtractFirstAndLastDigits extracts the first and last digits from a string.
func ExtractFirstAndLastDigits(s string) (firstDigit rune, lastDigit rune, err error) {
	// Remove leading and trailing whitespace
	s = strings.TrimSpace(s)

	// Check if the string is empty after removing whitespace
	if len(s) == 0 {
		return 0, 0, errors.New("input string is empty")
	}

	// Find the first digit
	for _, r := range s {
		if '0' <= r && r <= '9' {
			firstDigit = r
			break
		}
	}

	// If no digits were found, return an error
	if firstDigit == 0 {
		return 0, 0, errors.New("no digits found in the input string")
	}

	// Find the last digit
	lastDigit = 0
	for i := len(s) - 1; i >= 0; i-- {
		r := rune(s[i])
		if '0' <= r && r <= '9' {
			lastDigit = r
			break
		}
	}

	return
}

func readFileLines(filename string) ([]string, error) {
	var lines []string
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	return lines, nil
}

func getCaliValFromAll(filename string) {
	lines, err := readFileLines(filename)
	if err != nil {
		fmt.Println(err)
		return
	}

	// Create a dictionary (map) of strings to integers
	stringToIntMap := make(map[string]string)
	stringToIntMap["one"] = "1"
	stringToIntMap["two"] = "2"
	stringToIntMap["three"] = "3"
	stringToIntMap["four"] = "4"
	stringToIntMap["five"] = "5"
	stringToIntMap["six"] = "6"
	stringToIntMap["seven"] = "7"
	stringToIntMap["eight"] = "8"
	stringToIntMap["nine"] = "9"

	total := 0

	for _, line := range lines {
		nonOverlapLine := line
		for k := range stringToIntMap {
			nonOverlapLine = strings.ReplaceAll(nonOverlapLine, k, k+string(k[len(k)-1]))
		}
		for k := range stringToIntMap {
			nonOverlapLine = strings.ReplaceAll(nonOverlapLine, k, stringToIntMap[k])
		}

		firstDigit, lastDigit, err := ExtractFirstAndLastDigits(nonOverlapLine)
		if err != nil {
			fmt.Println(err)
		} else {
			value := int(firstDigit-'0')*10 + int(lastDigit-'0')
			total += value
		}
	}

	fmt.Printf("%s Part 2 Calibration = %d\n", filename, total)
}

func getCalibrationValue(filename string) {
	// filename := "example.dat"
	lines, err := readFileLines(filename)
	if err != nil {
		fmt.Println(err)
		return
	}

	total := 0

	for _, line := range lines {
		// fmt.Println(line)
		firstDigit, lastDigit, err := ExtractFirstAndLastDigits(line)
		if err != nil {
			fmt.Println(err)
		} else {
			value := int(firstDigit-'0')*10 + int(lastDigit-'0')
			// fmt.Printf("Digits: %c %c %d\n", firstDigit, lastDigit, value)
			// fmt.Printf("Last Digit: %c\n", lastDigit)
			total += value
		}
	}

	fmt.Printf("%s Calibration = %d\n", filename, total)
}

func main() {
	getCalibrationValue("example.dat")
	getCalibrationValue("data.dat")
	getCaliValFromAll("example2.dat")
	getCaliValFromAll("data.dat")
}
